File: backbone-go/examples/clean-api-go/application/usecases/get_products.go
// Package usecases contains application use cases
package usecases

import (
	"context"
	"time"

	"github.com/freakjazz/backbone-go/application/exceptions"
	"github.com/freakjazz/backbone-go/domain/specifications"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

// GetProductsInput carries the 4 generic query parameters.
//
//	Filters  — each string: "field,operator,value[,condition]"
//	           operators : eq ne gt gte lt lte contains in between is_null is_not_null
//	           conditions: and (default) | or
//	           examples  : "category,eq,Electronics,and"  "price,gt,500,and"
//	Page     — page number (default 1)
//	PageSize — items per page (default 10)
//	SortBy   — "field:direction" e.g. "price:desc" (default "created_at:desc")
type GetProductsInput struct {
	Filters  []string `json:"filters"`
	Page     int      `json:"page"`
	PageSize int      `json:"page_size"`
	SortBy   string   `json:"sort_by"`
}

// GetProductsOutput is the result of the use case.
type GetProductsOutput struct {
	Products   []*entities.Product
	TotalCount int64
	Page       int
	PageSize   int
}

// GetProductsUseCase handles listing products with dynamic filters.
type GetProductsUseCase struct {
	repository repositories.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewGetProductsUseCase creates a new use case instance.
func NewGetProductsUseCase(repository repositories.ProductRepository, logger *logging.EnhancedLogger) *GetProductsUseCase {
	return &GetProductsUseCase{repository: repository, logger: logger}
}

// Execute applies filters, sorts and pagination via the Specification + Criteria pattern.
func (uc *GetProductsUseCase) Execute(ctx context.Context, input GetProductsInput) (*GetProductsOutput, error) {
	ucLogger := uc.logger.
		WithLayer("application").
		WithHandler("GetProductsUseCase").
		WithMethod("Execute").
		WithContext(map[string]interface{}{
			"request_id": ctx.Value("request_id"),
			"filters":    input.Filters,
			"page":       input.Page,
			"page_size":  input.PageSize,
			"sort_by":    input.SortBy,
		})

	ucLogger.Info("Getting products", nil)

	// Defaults
	page := input.Page
	if page < 1 {
		page = 1
	}
	pageSize := input.PageSize
	if pageSize < 1 {
		pageSize = 10
	}
	if pageSize > 100 {
		pageSize = 100
	}

	sortField, sortDir := specifications.ParseSortBy(input.SortBy)

	// Build Criteria from generic filter params
	start := time.Now()
	criteria := specifications.ParseFilterParams(input.Filters, page, pageSize, sortField, sortDir)

	sql, args := criteria.GetFullSQL("SELECT * FROM products")
	ucLogger.Debug("Criteria built", map[string]interface{}{
		"sql":            sql,
		"args":           args,
		"filters_count":  len(input.Filters),
	})

	// Count
	totalCount, err := uc.repository.Count(ctx, criteria)
	if err != nil {
		ucLogger.ErrorWithCode("Failed to count products", 10002001, map[string]interface{}{"error": err.Error()})
		return nil, exceptions.NewUseCaseException("Failed to count products", "GetProductsUseCase")
	}

	// Fetch
	products, err := uc.repository.FindByCriteria(ctx, criteria)
	if err != nil {
		ucLogger.ErrorWithCode("Failed to get products", 10002002, map[string]interface{}{
			"error":       err.Error(),
			"duration_ms": time.Since(start).Milliseconds(),
		})
		return nil, exceptions.NewUseCaseException("Failed to get products", "GetProductsUseCase")
	}

	ucLogger.Info("Products retrieved", map[string]interface{}{
		"count":       len(products),
		"total":       totalCount,
		"duration_ms": time.Since(start).Milliseconds(),
	})

	return &GetProductsOutput{
		Products:   products,
		TotalCount: totalCount,
		Page:       page,
		PageSize:   pageSize,
	}, nil
}

File: examples/clean_api_python/application/use_cases/get_products.py
"""Get products use case - Application layer"""
from dataclasses import dataclass
from typing import List, Dict, Any, Optional
import sys
import os

# Add backbone to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..')))

from backbone.domain.specifications import FilterSpecification, SortSpecification
from backbone.infrastructure.logging import LoggerFactory, LogLevel
from examples.clean_api_python.domain.entities.product import Product
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository
from examples.clean_api_python.domain.specifications import product_specs
import time


@dataclass
class GetProductsInput:
    """Input for getting products with filters"""
    category: Optional[str] = None
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    in_stock: bool = False
    active: bool = True
    name_pattern: Optional[str] = None
    page: int = 1
    page_size: int = 10
    sort_by: str = "created_at"
    sort_order: str = "desc"


@dataclass
class GetProductsOutput:
    """Output of getting products"""
    products: List[Product]
    total_count: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_previous: bool


class GetProductsUseCase:
    """Handles getting products with filters"""
    
    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api",
            "application",
            "GetProductsUseCase"
        )
    
    def execute(self, input: GetProductsInput, context: Dict[str, Any]) -> GetProductsOutput:
        """Executes the use case with Specification Pattern"""
        self.logger.info(
            "Getting products with filters",
            extra_data={
                "category": input.category,
                "min_price": input.min_price,
                "max_price": input.max_price,
                "in_stock": input.in_stock,
                "page": input.page,
                "page_size": input.page_size
            }
        )
        
        # Construir filtros con Specification Pattern
        start = time.time()
        filters = self._build_filters(input)
        sorts = self._build_sorts(input)
        
        self.logger.debug(
            "Query criteria built",
            extra_data={
                "filters_applied": self._get_applied_filters(input),
                "filters_count": len(filters)
            }
        )
        
        # Obtener total count
        total_count = self.repository.count(filters)
        
        # Obtener productos
        try:
            products = self.repository.find_by_criteria(
                filters=filters,
                sorts=sorts,
                page=input.page,
                page_size=input.page_size
            )
        except Exception as e:
            duration_ms = int((time.time() - start) * 1000)
            self.logger.error(
                "Failed to get products",
                extra_data={
                    "error": str(e),
                    "duration_ms": duration_ms,
                    "error_code": 10002002
                }
            )
            raise Exception(f"Failed to get products: {str(e)}")
        
        duration_ms = int((time.time() - start) * 1000)
        
        # Calcular paginación
        total_pages = (total_count + input.page_size - 1) // input.page_size
        
        output = GetProductsOutput(
            products=products,
            total_count=total_count,
            page=input.page,
            page_size=input.page_size,
            total_pages=total_pages,
            has_next=input.page < total_pages,
            has_previous=input.page > 1
        )
        
        self.logger.info(
            "Products retrieved successfully",
            extra_data={
                "count": len(products),
                "total_count": total_count,
                "duration_ms": duration_ms
            }
        )
        
        return output
    
    def _build_filters(self, input: GetProductsInput) -> List[FilterSpecification]:
        """Builds filters with Specification Pattern"""
        filters = []
        
        # Filtro de activo por defecto
        if input.active:
            filters.append(product_specs.product_is_active())
        
        # Filtro por categoría
        if input.category:
            filters.append(product_specs.product_in_category(input.category))
        
        # Filtro por rango de precio
        if input.min_price and input.max_price:
            filters.append(product_specs.product_price_range(input.min_price, input.max_price))
        elif input.min_price:
            filters.append(FilterSpecification("price", ">=", input.min_price))
        elif input.max_price:
            filters.append(FilterSpecification("price", "<=", input.max_price))
        
        # Filtro por stock
        if input.in_stock:
            filters.append(product_specs.product_in_stock())
        
        # Filtro por nombre (pattern)
        if input.name_pattern:
            filters.append(product_specs.product_by_name_pattern(input.name_pattern))
        
        return filters
    
    def _build_sorts(self, input: GetProductsInput) -> List[SortSpecification]:
        """Builds sorting"""
        return [SortSpecification(input.sort_by, input.sort_order)]
    
    def _get_applied_filters(self, input: GetProductsInput) -> List[str]:
        """Returns applied filters for logging"""
        filters = []
        if input.category:
            filters.append("category")
        if input.min_price or input.max_price:
            filters.append("price_range")
        if input.in_stock:
            filters.append("in_stock")
        if input.name_pattern:
            filters.append("name_pattern")
        return filters
